internal/provider/resource_gitlab_project_issue.go

```diff
@@ -0,0 +1,294 @@
+				},
+			},
+		),
+	}
+})
+
+func resourceGitlabProjectIssueCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+
+	options := &gitlab.CreateIssueOptions{
+		Title: gitlab.String(d.Get("title").(string)),
+	}
+	if iid, ok := d.GetOk("iid"); ok {
+		options.IID = gitlab.Int(iid.(int))
+	}
+	if assigneeIDs, ok := d.GetOk("assignee_ids"); ok {
+		options.AssigneeIDs = intSetToIntSlice(assigneeIDs.(*schema.Set))
+	}
+	if confidential, ok := d.GetOk("confidential"); ok {
+		options.Confidential = gitlab.Bool(confidential.(bool))
+	}
+	if createdAt, ok := d.GetOk("created_at"); ok {
+		parsedCreatedAt, err := time.Parse(time.RFC3339, createdAt.(string))
+		if err != nil {
+			return diag.Errorf("failed to parse created_at: %s. It must be in valid RFC3339 format.", err)
+		}
+		options.CreatedAt = gitlab.Time(parsedCreatedAt)
+	}
+	if description, ok := d.GetOk("description"); ok {
+		options.Description = gitlab.String(description.(string))
+	}
+	if discussionToResolve, ok := d.GetOk("discussion_to_resolve"); ok {
+		options.DiscussionToResolve = gitlab.String(discussionToResolve.(string))
+	}
+	if dueDate, ok := d.GetOk("due_date"); ok {
+		parsedDueDate, err := parseISO8601Date(dueDate.(string))
+		if err != nil {
+			return diag.Errorf("failed to parse due_date: %s. %v", dueDate.(string), err)
+		}
+		options.DueDate = parsedDueDate
+	}
+	if issueType, ok := d.GetOk("issue_type"); ok {
+		options.IssueType = gitlab.String(issueType.(string))
+	}
+	if labels, ok := d.GetOk("labels"); ok {
+		gitlabLabels := gitlab.Labels(*stringSetToStringSlice(labels.(*schema.Set)))
+		options.Labels = &gitlabLabels
+	}
+	if mergeRequestToResolveDiscussionsOf, ok := d.GetOk("merge_request_to_resolve_discussions_of"); ok {
+		options.MergeRequestToResolveDiscussionsOf = gitlab.Int(mergeRequestToResolveDiscussionsOf.(int))
+	}
+	if milestoneID, ok := d.GetOk("milestone_id"); ok {
+		options.MilestoneID = gitlab.Int(milestoneID.(int))
+	}
+	if weight, ok := d.GetOk("weight"); ok {
+		options.Weight = gitlab.Int(weight.(int))
+	}
+
+	issue, _, err := client.Issues.CreateIssue(project, options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	updateOptions := gitlab.UpdateIssueOptions{}
+	if discussionLocked, ok := d.GetOk("discussion_locked"); ok {
+		updateOptions.DiscussionLocked = gitlab.Bool(discussionLocked.(bool))
+	}
+	if stateEvent, ok := d.GetOk("state"); ok {
+		updateOptions.StateEvent = gitlab.String(issueStateToStateEvent[stateEvent.(string)])
+	}
+	if updateOptions != (gitlab.UpdateIssueOptions{}) {
+		_, _, err := client.Issues.UpdateIssue(project, issue.IID, &updateOptions, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.Errorf("failed to update issue %d in project %s right after creation: %v", issue.IID, project, err)
+		}
+	}
+
+	d.SetId(resourceGitLabProjectIssueBuildId(project, issue.IID))
```

Comment:
Always set the ID as soon as you can. In this case, after the `client.Issues.CreateIssue` call. The reason is it makes the creation idempotent if the second API call to update the issue fails. As the code is written now, if the update call fails, then the whole resource will not be tracked in state, and the user would need to either import it or delete it manually.